core/src/main/clojure/xtdb/types.clj

```diff
@@ -785,6 +785,48 @@
 (defn trunc-duration-to-micros [^Duration dur]
   (.withNanos dur (* (quot (.getNano dur) 1000) 1000)))
 
+(def pg-ranges
+  ;; we only support one
+  {:tzts-range
+   {:rngtypid 3910
+    :rngsubtype 1184
+    :rngmultitypid 4534
+    :rngcollation 0
+    :rngsubopc 3127
+    :rngcanonical ""
+    :rngsubdiff "tstzrange_subdiff"}})
+
+(defn binary-list-header [length element-oid typlen]
+  ;; list: <<byte-length,dimensions=1,data-offset=0,oid,elem-count=(lenght-of-list),l-bound=1?,data>>
+  ;; https://github.com/postgres/postgres/blob/master/src/include/utils/array.h
+  (let [payload-size (+ (* 5 4) (* typlen length))] ;; 5 integers describing array + data
+    (doto
+     (ByteBuffer/allocate (+ 4 payload-size)) ;; 4 is for int4 - payload size
+      (.putInt payload-size)
+      (.putInt 1) ;; dimensions
+      (.putInt 0) ;; data offset
+      (.putInt element-oid)
+      (.putInt length)
+      (.putInt 1) ;; l-bound - MUST BE 1
+      )))
+
+(defn read-binary-array [ba]
+  (let [bb (ByteBuffer/wrap ba)
+        ;; _payload-size (.getInt bb)
+        ;; _dimensions (.getInt bb 4)
+        data-offset (.getInt bb 8)
+        oid (.getInt bb 12)
+        length (.getInt bb 16)
+        [typlen get-elem] (case oid
+                            (23 1007) [4 (fn [bb idx] (.getInt bb idx))]
+                            (20 1016) [8 (fn [bb idx] (.getLong bb idx))])
+        start  (+ data-offset 20)]
+    (persistent!
```

Comment:
nit: think this can be `(into [] (map (fn [idx] (get-elem ...)) (range length))`, which will do the reduce/conj, transient/persistent dances for you?